clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -558,15 +558,35 @@ static void collectSYCLAttributes(Sema &S, FunctionDecl *FD,
   if (!FD->hasAttrs())
     return;
 
-  llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
-    // FIXME: Make this list self-adapt as new SYCL attributes are added.
-    return isa<IntelReqdSubGroupSizeAttr, IntelNamedSubGroupSizeAttr,
-               ReqdWorkGroupSizeAttr, SYCLIntelKernelArgsRestrictAttr,
-               SYCLIntelNumSimdWorkItemsAttr,
-               SYCLIntelSchedulerTargetFmaxMhzAttr,
-               SYCLIntelMaxWorkGroupSizeAttr, SYCLIntelMaxGlobalWorkDimAttr,
-               SYCLIntelNoGlobalWorkOffsetAttr, SYCLSimdAttr>(A);
-  });
+  // Attributes that should be propagated from device functions to a kernel
+  // in SYCL 1.2.1.
+  if (S.getASTContext().getLangOpts().getSYCLVersion() <
+      LangOptions::SYCL_2020) {
+    llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
+      // FIXME: Make this list self-adapt as new SYCL attributes are added.
+      return isa<IntelReqdSubGroupSizeAttr, IntelNamedSubGroupSizeAttr,
+                 ReqdWorkGroupSizeAttr, SYCLIntelKernelArgsRestrictAttr,
+                 SYCLIntelNumSimdWorkItemsAttr,
+                 SYCLIntelSchedulerTargetFmaxMhzAttr,
+                 SYCLIntelMaxWorkGroupSizeAttr, SYCLIntelMaxGlobalWorkDimAttr,
+                 SYCLIntelNoGlobalWorkOffsetAttr, SYCLSimdAttr>(A);
+    });
+  } else {
+    // Attributes that should not be propagated from device functions to a
+    // kernel in SYCL 2020.
+    if (DirectlyCalled) {
+      llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
+        return isa<
+            SYCLIntelFPGAMaxConcurrencyAttr,
+            SYCLIntelFPGADisableLoopPipeliningAttr, SYCLSimdAttr,
+            SYCLIntelKernelArgsRestrictAttr, ReqdWorkGroupSizeAttr,
+            SYCLIntelNumSimdWorkItemsAttr, SYCLIntelSchedulerTargetFmaxMhzAttr,
+            SYCLIntelNoGlobalWorkOffsetAttr, SYCLIntelMaxWorkGroupSizeAttr,
+            IntelReqdSubGroupSizeAttr, SYCLIntelMaxGlobalWorkDimAttr,
+            IntelNamedSubGroupSizeAttr, SYCLIntelFPGAInitiationIntervalAttr>(A);
+      });
+    }
+  }
 
   // Attributes that should not be propagated from device functions to a kernel.
   if (DirectlyCalled) {
```

Comment:
Some of these attributes are being added twice for SYCL 2020. E.g. `SYCLIntelFPGAMaxConcurrencyAttr`. Some like  `SYCLIntelLoopFuseAttr` is missed in SYCL2020.

Reply:
> Some of these attributes are being added twice for SYCL 2020. E.g. `SYCLIntelFPGAMaxConcurrencyAttr`.

My mistake. Fixed now.

> Some like `SYCLIntelLoopFuseAttr` is missed in SYCL2020.

I did not make any change in this list  below because all attributes will be applied both in SYCL 2017 and SYCL 2020 modes when we call DirectlyCalled.

```
// Attributes that should not be propagated from device functions to a kernel.
  if (DirectlyCalled) {
    llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
      return isa<SYCLIntelLoopFuseAttr, SYCLIntelFPGAMaxConcurrencyAttr,
                 SYCLIntelFPGADisableLoopPipeliningAttr,
                 SYCLIntelFPGAInitiationIntervalAttr,
                 SYCLIntelUseStallEnableClustersAttr>(A);
    });
```

Reply:
Given that these lists are identical now.. do we want to combine the logic into one:
```
if (DirectlyCalled || S.getASTContext().getLangOpts().getSYCLVersion() < LangOptions::SYCL_2020) {
  llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
        return isa<IntelReqdSubGroupSizeAttr, IntelNamedSubGroupSizeAttr,
                   ReqdWorkGroupSizeAttr, SYCLIntelKernelArgsRestrictAttr,
                   SYCLIntelNumSimdWorkItemsAttr,
                   SYCLIntelSchedulerTargetFmaxMhzAttr,
                   SYCLIntelMaxWorkGroupSizeAttr, SYCLIntelMaxGlobalWorkDimAttr,
                   SYCLIntelNoGlobalWorkOffsetAttr, SYCLSimdAttr>(A);
      });
}
```

Reply:
@elizabethandrews, are you OK with the change? Do you have any questions?

Reply:
The code change LGTM. I haven't had a chance to look over test changes. I will take a look ASAP